libmscore/changeMap.h

```diff
@@ -0,0 +1,103 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2009-2019 Werner Schweer
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#ifndef __CHANGEMAP_H__
+#define __CHANGEMAP_H__
+
+#include "fraction.h"
+
+/**
+ \file
+ Definition of class ChangeMap.
+*/
+
+namespace Ms {
+
+enum class ChangeMethod : signed char {
+      NORMAL,
+      EXPONENTIAL,
+      EASE_IN,
+      EASE_OUT,
+      EASE_IN_OUT       // and shake it all about
+      };
+
+enum class ChangeDirection : signed char {
+      INCREASING,
+      DECREASING
+      };
+
+//---------------------------------------------------------
+///   ChangeEvent
+///   item in ChangeMap
+//---------------------------------------------------------
+
+enum class ChangeEventType : char { FIX, RAMP };
+
+class ChangeEvent {
+      // Despite storing the tick as the key of the ChangeEvent in the ChangeMap, we also store it here.
+      // Since we're not going to be changing it (well, not much, anyway), this is good because it provides
+      // multiple ways to access the tick. Keeping it in sync isn't much trouble.
+      Fraction tick;
```

Comment:
Looking through this code, I would rather suggest not binding the event structure itself to absolute tick values but rather store only its length within the event itself and rely on the values in `ChangeMap` for information about an absolute tick position. I have briefly searched for this tick usage (when it was still `VeloMap`), and, although I could have missed something, it looks like you use it for events obtained from VeloMap by tick or from iterator, so finding tick with this approach shouldn't be problematic. On the other side, not binding an event to a tick would simplify handling events in other contexts, I should probably mention it somewhere in other comments to this PR.

Reply:
Right, I thought so too. So I started working on replacing it with just a length and storing the start tick only as the key of the map. Then I realised one of the reasons why I didn't do that in the first place:

In this loop, for example:

```cpp
      //
      // Stage 2: readjust lengths of any colliding ramps:
      //
      int j = -1;
      for (auto& event : values()) {
            if (event.type != ChangeEventType::RAMP)
                  continue;

            j++;
            if (!startsInRamp[j])
                  continue;

            // Take a copy of the event...
            auto tempChangeEvent = event;
            remove(event.tick, event);
            tempChangeEvent.tick = endPoints[j-1].second;
            // ...and move it to its new tick position
            insert(tempChangeEvent.tick, tempChangeEvent);
            }
```

If I iterate using `values()` it works fine, because it's a copy of the values in the array. However, with an iterator it's a lot more difficult, because I can't modify the map while iterating over it with an iterator. I'd have to somehow store the operations that I need to do (i.e. moving and changing length) and then have another loop to actually carry out those operations. It just all becomes a bit difficult.

I can still try to implement it, but I don't know how easy it will be. Thoughts?

Reply:
This issue should be easy to resolve exactly because it is a map. Similarly to STL maps, inserting or removing items to/from map does not invalidate all iterators in it:

> If you add items to the map, existing iterators will remain valid. If you remove items from the map, iterators that point to the removed items will become dangling iterators.

(from [this page](https://doc.qt.io/qt-5/qmap-iterator.html))

So the only thing that should be ensured is that you obtain the next iterator in the map *before* erasing elements from it rather than after it. The [same page](https://doc.qt.io/qt-5/qmap-iterator.html) from Qt docs provides some examples of this:
```
QMap<QString, int>::iterator i = map.begin();
while (i != map.end()) {
    if (i.key().startsWith('_'))
        i = map.erase(i);
    else
        ++i;
}
```
```
QMap<QString, int>::iterator i = map.begin();
while (i != map.end()) {
    QMap<QString, int>::iterator prev = i;
    ++i;
    if (prev.key().startsWith('_'))
        map.erase(prev);
}
```

Something similar should probably work in this case as well.

Reply:
Sounds good, thank you!